programs/mmm/src/instructions/mpl_core_asset/mpl_core_fulfill_buy.rs

```diff
@@ -0,0 +1,376 @@
+use anchor_lang::{prelude::*, AnchorDeserialize};
+use mpl_core::{instructions::TransferV1Builder, types::UpdateAuthority};
+use solana_program::program::invoke;
+use std::convert::TryFrom;
+
+use crate::{
+    assert_valid_core_plugins,
+    constants::*,
+    deserialize_collection_asset,
+    errors::MMMErrorCode,
+    get_royalties_from_plugin, index_ra,
+    instructions::{
+        check_allowlists_for_mpl_core, check_remaining_accounts_for_m2, create_core_metadata_core,
+        withdraw_m2,
+    },
+    state::{Pool, SellState},
+    util::{
+        assert_valid_fees_bp, get_buyside_seller_receives, get_lp_fee_bp, get_metadata_royalty_bp,
+        get_sol_fee, get_sol_lp_fee, get_sol_total_price_and_next_price, log_pool,
+        pay_creator_fees_in_sol, try_close_escrow, try_close_pool, try_close_sell_state,
+    },
+    AssetInterface, IndexableAsset, SolFulfillBuyArgs,
+};
+
+#[derive(Accounts)]
+#[instruction(args:SolFulfillBuyArgs)]
+pub struct MplCoreFulfillBuy<'info> {
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    /// CHECK: we will check the owner field that matches the pool owner
+    #[account(mut)]
+    pub owner: UncheckedAccount<'info>,
+    pub cosigner: Signer<'info>,
+    #[account(mut)]
+    /// CHECK: we will check that the referral matches the pool's referral
+    pub referral: UncheckedAccount<'info>,
+    #[account(
+        mut,
+        seeds = [POOL_PREFIX.as_bytes(), owner.key().as_ref(), pool.uuid.as_ref()],
+        has_one = owner @ MMMErrorCode::InvalidOwner,
+        has_one = referral @ MMMErrorCode::InvalidReferral,
+        has_one = cosigner @ MMMErrorCode::InvalidCosigner,
+        constraint = pool.payment_mint.eq(&Pubkey::default()) @ MMMErrorCode::InvalidPaymentMint,
+        constraint = pool.expiry == 0 || pool.expiry > Clock::get().unwrap().unix_timestamp @ MMMErrorCode::Expired,
+        bump
+    )]
+    pub pool: Box<Account<'info, Pool>>,
+    /// CHECK: it's a pda, and the private key is owned by the seeds
+    #[account(
+        mut,
+        seeds = [BUYSIDE_SOL_ESCROW_ACCOUNT_PREFIX.as_bytes(), pool.key().as_ref()],
+        bump,
+    )]
+    pub buyside_sol_escrow_account: UncheckedAccount<'info>,
+    #[account(
+        mut,
+        constraint = asset.to_account_info().owner == asset_program.key,
```

Comment:
I believe this is unnecessary, the `anchor_lang::Owner` trait is already implemented